### DISCUSSION TITLE
36 sound inits at launch time

### DISCUSSION THREAD

**Discussion by ngocdh** [jamulussoftware/jamulus#1727] [Bug Investigation]
While investigating iOS (#1512) slow launch time, I and found that Client.init() is called 9 times at launch, each Client.init() calls 4 Sound.init(), which makes 36 Sound inits at launch time. 

For iOS, one init is enough, so I made a commit for iOS (in #1512) to do just that. We might need to improve for other OSes so that Jamulus launches faster.

**Comment by hoffie** [comment]
The the first time I'm wondering: Should this rather be an Issue to track it than a Discussion? :)
It seems actionable and needs tracking. @gilgongo What do you think?

**Reply by gilgongo** [reply]
Yes if it's a clear description of a bug or otherwise an actionable piece of work.

**Comment by dcorson-ticino-com** [comment]
How do you do the supported buffer size requests with only 1 call to Init?
How do you know when the setup requests are finished to send a final Init?
I have not looked at the timing and sequencing, but only when the sequencing is known can we start to optimize.
Or is the iOS audio interface so different?
I'm not doubtful, just wondering.

**Reply by ngocdh** [reply]
For iOS, I believe its simpler because all iOS devices should be similar. Unlike Windows, Mac, or Linux, etc. or even Android which can have different/multiple soundcards.

I've been testing iOS for a while, and my devices (iphone 12 and ipad pro) support all three buffer sizes (64, 128, 256). Other parameters are irrelevant to iOS (quality/opus, left/right channels, etc.). So I went ahead and remove all those capabilities checking for iOS.  (I created a bug when doing this, but it's unrelated, fixing it)

Don't know about other OSes though.

**Reply by dcorson-ticino-com** [reply]
So that means it is impossible to use an external sound card using iOS ?

If there is any way that an external sound card could possibly be used with iOS then I think we must keep all the possibilities to configure it.

**Reply by ngocdh** [reply]
Right! I didn't think of external sound card. 
Problem I'm trying to solve is 36 inits take long on iOS, and because of that @ann0see 's iPhone SE couldn't load it fast enough the app gets killed after like 10s.

Still, do you think we can at least lower the number of inits down? 36 seems a lot to me.

**Reply by ngocdh** [reply]
So I did a test with external sound card on iOS. As far as I could tell, iOS treated the external sound card no differently, meaning you get stereo sound, but one input channel only (the whole mix from the soundcard), just like the built-in device.
The only problem seems to be the check against buffer sizes (64, 128, 256). My soundcard worked fine with all 3 settings, but maybe there are others that won't.

Now if all the checks are really necessary, I suggest we do it using multi-threading so that Jamulus doesn't get blocked by the checks. Once Jamulus launched, users can still perform usual tasks, such as modifying profile settings, browsing rooms, etc. 
But only when all checks are done they can connect to a room.

**Reply by dcorson-ticino-com** [reply]
I have nothing against reducing the init calls, it should just not limit the usage.
I do not know, for example, if there is a fixed order of setup calls which would allow using just one call at the end.
A detailed reading of the code is necessary.

**Comment by ngocdh** [comment]
Idea: we don't need to run the capability checks every time. We can save results from the first check round and load them at launch. If soundcard is the same, then no need to rerun the checks. So 36 sound inits down to 1 only.

What do you think?

**Reply by ann0see** [reply]
Not sure if that solves the problem. On Windows the driver Name might stay the same (ASIO4ALL) but in reality it‘s a totally different device.

**Reply by ngocdh** [reply]
Then make an exception. At least mobile devices would benefit much from this.

**Reply by npostavs** [reply]
For ASIO in particular, the API can tell the range of buffer sizes that the device will work with. So you don't need to test each one with `Init()`.

**Reply by ngocdh** [reply]
That seems to be good news, thanks @npostavs!
